118 patch method

ksuderman opened the pull request:
Adds support for the HTTP PATCH method.  See #118 

This PR is for comments and discussion and may need further changes before being merged.

cjstehno commented on build.gradle:
```diff
@@ -8,7 +8,7 @@ plugins {
 }
 
 group = 'io.github.http-builder-ng'
-version = '0.15.0'
```
Just FYI: We don't do SNAPSHOTS, so please remove this before the final merge.

cjstehno commented on build.gradle:
```diff
@@ -8,7 +8,7 @@ plugins {
 
 group = 'io.github.http-builder-ng'
-version = '0.15.0'
+version = '0.16.0-SNAPSHOT'
```
Just FYI: We don't do SNAPSHOTS, so please remove this before the final merge.

cjstehno commented on http-builder-ng-apache/src/test/groovy/groovyx/net/http/ApacheHttpPatchSpec.groovy:
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2017 David Clark
```
I will have an update to the copyright headers going in shortly... changing it to "HttpBuilder-NG Project" rather than Dave specifically (he approved). Sorry, but this may result in some conflicts that you will need to resolve on your end.

cjstehno commented on http-builder-ng-core/src/main/java/groovyx/net/http/HttpBuilder.java:
```diff
@@ -1506,6 +1506,239 @@ public Object delete(final Consumer<HttpConfig> configuration) {
+    public Object patch(final Consumer<HttpConfig> configuration) {
+        return patch(Object.class, configuration);
+    }
+
```
Nitpicky I know, but please remove the extra line here.

cjstehno commented on http-builder-ng-core/src/main/java/groovyx/net/http/JavaHttpBuilder.java:
```diff
@@ -323,6 +323,16 @@ protected Object doDelete(final ChainedHttpConfig requestConfig) {
+        // A hack-around is to set header "X-HTTP-Method-Override=PATCH" and
+        // send a POST message instead.  It looks like this would be done around
+        // line 58 above.
+        return createAndExecute(requestConfig, "PATCH");
```
As discussed in the issue thread, go ahead and throw some sort of `RuntimeException` here. The core PATCH test will just be a standalone test to ensure that the exception is thrown (disconnected from the test kit).

After this gets merged into `development` I will update the User Guide to document that PATCH is not supported in the core implementation and why.

cjstehno commented on http-builder-ng-core/src/test/groovy/groovyx/net/http/JavaHttpPatchSpec.groovy:
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright (C) 2017 Keith Suderman
```
This should be changed to "HttpBuilder-NG Project" - as all of the others will be shortly.

cjstehno commented on http-builder-ng-core/src/test/groovy/groovyx/net/http/JavaHttpPatchSpec.groovy:
```diff
@@ -0,0 +1,23 @@
+import groovyx.net.http.tk.HttpPutTestKit
+
+/**
+ * @author Keith Suderman
```
We don't use the `author` tags and prefer to use `git blame` to track authorship as needed. I am going to start a "contributors" section of the User Guide for making note of our recent contributors (with a link to their GitHub profile).

cjstehno commented on http-builder-ng-core/src/test/groovy/groovyx/net/http/tk/HttpPatchTestKit.groovy:
```diff
@@ -0,0 +1,625 @@
+/*
+ * Copyright (C) 2017 Keith Suderman
```
Realize this is probably a template, but as above this should be "HttpBuilder-NG Project".

ksuderman commented:
I believe I have resolved all of the above. 

1. Fixed copyright headers
1. Removed @author tag
1. Remove extraneous blank line in HttpBuilder.java
1. JavaHttpBuilder.doPatch() throws an UnsupportedOperationException
    - message is "java.net.HttpURLConnection does not support the PATCH method. Use the Apache or OkHttp providers instead."
1. JavaHttpPathSpec checks to ensure the proper exception is thrown from:
    1. patch()
    1. patchAsync()
    1. patch(Consumer<HttpConfig>)
    1. patchAsync(Consumer<HttpConfig>)

The four tests *should* exercise all the possible paths to `HttpUrlConnection.setRequestMethod`.

cjstehno commented:
Great. I will review tomorrow morning. Thanks​.

cjstehno commented:
Merged in... thanks for the contribution. The 0.16 release should be cut within the next couple days.
